Adapt platform API access to CAPI (#2229)
giantswarm/docs: src/content/overview/fleet-management/multi-tenancy/_index.md

---
title: Multi-tenancy
description: Manage cluster and application resources across multiple organizations and teams.
weight: 40
menu:
  principal:
    parent: overview-fleet-management
    identifier: overview-fleet-management-multi-tenancy

> last_review_date: 2024-06-11

Author: ```suggestion
last_review_date: 2024-07-17
```

```text
$ kubectl gs template organization --name e-commerce
apiVersion: security.giantswarm.io/v1alpha1
kind: Organization
metadata:
 name: e-commerce
# [...]
```

Once an organization CR is created, our automation ([organization-operator](https://github.com/giantswarm/organization-operator), to be precise) ensures that a namespace `org-<name of organization>` exists for the organization.

## Naming conventions {#naming-conventions}

Organization names (technically, organization CR names) must follow these rules:

- Must be unique within the management cluster
- Must contain at most 59 characters.
- Has to conform to the same convention as Kubernetes namespaces additionally (i. e. the [DNS label names convention](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)), which means:
    - contain only lowercase alphanumeric characters or '-'
    - start with an alphanumeric character
    - end with an alphanumeric character

Since each organization has a namespace created prefixed with `org-`, we recommend against using the same `org-` prefix in the organization name to avoid confusion.

## Organization namespace {#namespace}

> The management cluster creates a namespace for each organization. The namespace is named after the organization CR name, prefixed with `org—'.

Reviewer: ```suggestion
For each organization there exists a namespace in the management cluster. The namespace is named after the organization CR name, prefixed with `org—'.
```

## Organization CRD and CRs {#organization-crd-cr}

If the concept of custom resources (CR) and custom resource definitions (CRD) is new to you, Kubernetes allows you to define [arbitrary objects](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) handled via the Kubernetes API. A custom resource definition specifies the schema of such an object. The actual objects are called the custom resources.

Giant Swarm management clusters provide a CRD named `Organization` (long form: `organizations.security.giantswarm.io`, [schema documentation]({{< relref "/vintage/use-the-api/management-api/crd/organizations.security.giantswarm.io.md" >}})). An organization is defined simply by a custom resource using that CRD, which we'll call an "organization CR" here for brevity.

The single most crucial aspect of an organization's CR is its name. Therefore, the CR looks as simple as this:

```text
$ kubectl gs template organization --name e-commerce
apiVersion: security.giantswarm.io/v1alpha1
kind: Organization
metadata:
 name: e-commerce
# [...]
```

> Once an organization CR is created, our automation ([organization-operator](https://github.com/giantswarm/organization-operator), to be precise) ensures that a namespace `org-<name of organization>` exists for the organization.

Reviewer: ```suggestion
Once an organization CR is created, our custom operator called [organization-operator](https://github.com/giantswarm/organization-operator) ensures that a namespace `org-<name of organization>` exists for the organization.
```
